core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -220,9 +233,15 @@ protected boolean addsDeleteFiles() {
   /** Add a data file to the new snapshot. */
   protected void add(DataFile file) {
     Preconditions.checkNotNull(file, "Invalid data file: null");
-    setDataSpec(file);
-    addedFilesSummary.addedFile(dataSpec(), file);
+
+    PartitionSpec dataSpec = ops.current().spec(file.specId());
+    Preconditions.checkNotNull(
+        dataSpec, "Cannot find partition spec for data file: %s", file.path());
```

Comment:
Could we include the partition spec ID in the failure message? Also could we make this `Preconditions.checkArgument(dataSpec != null)`? I know it was already a notNull earlier but that leads to unclear NPEs as opposed to this case which I think really is an illegal argument exception (a data file with an invalid spec)